src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                }
+                else
+                {
+                    // if (condition)
+                    //     consequence;
+                    // else 
+                    //     alternative
+                    //
+                    // becomes
+                    //
+                    // GotoIfFalse condition alt;
+                    // consequence
+                    // goto afterif;
+                    // alt:
+                    // alternative;
+                    // afterif:
+
+                    BasicBlock whenFalse = null;
+                    VisitConditionalBranch(operation.Condition, ref whenFalse, false);
+
+                    VisitStatement(operation.WhenTrue);
+
+                    var afterIf = new BasicBlock(BasicBlockKind.Block);
+                    LinkBlocks(CurrentBasicBlock, afterIf);
+                    _currentBasicBlock = null;
+
+                    AppendNewBlock(whenFalse);
+                    VisitStatement(operation.WhenFalse);
+
+                    AppendNewBlock(afterIf);
+                }
+
+                return null;
+            }
+            else
+            {
+                // condition ? consequence : alternative
+                //
+                // becomes
+                //
+                // GotoIfFalse condition alt;
+                // capture = consequence
+                // goto afterif;
+                // alt:
+                // capture = alternative;
+                // afterif:
+                // result - cupture
+
+                SpillEvalStack();
+                int captureId = _availableCaptureId++;
+
+                BasicBlock whenFalse = null;
+                VisitConditionalBranch(operation.Condition, ref whenFalse, false);
+
+                AddStatement(new SimpleAssignmentExpression(new FlowCapture(captureId, operation.Syntax, 
+                                                                            isInitialization:true, operation.Type, 
+                                                                            default(Optional<object>)),
+                                                            operation.IsRef,
+                                                            Visit(operation.WhenTrue),
+                                                            semanticModel:null,
+                                                            operation.WhenTrue.Syntax,
+                                                            operation.Type,
+                                                            default(Optional<object>),
+                                                            isImplicit:true));
+
+                var afterIf = new BasicBlock(BasicBlockKind.Block);
+                LinkBlocks(CurrentBasicBlock, afterIf);
+                _currentBasicBlock = null;
+
+                AppendNewBlock(whenFalse);
+
+                AddStatement(new SimpleAssignmentExpression(new FlowCapture(captureId, operation.Syntax,
```

Comment:
We're using the same `captureId` twice. Is that correct, or should it be incremented?
If ok to re-use, then should we re-use the `FlowCapture` instance instead?

Reply:
>We're using the same captureId twice. Is that correct, or should it be incremented?

It is  reference to the same capture, thus the same Id.

>If ok to re-use, then should we re-use the FlowCapture instance instead?

IOperation instances cannot be reused, they have parent pointers.


---
In reply to: [162177432](https://github.com/dotnet/roslyn/pull/24263#discussion_r162177432) [](ancestors = 162177432)